x/params/subspace/subspace.go

```diff
@@ -0,0 +1,198 @@
+	cdc  *codec.Codec
+	key  sdk.StoreKey // []byte -> []byte, stores parameter
+	tkey sdk.StoreKey // []byte -> bool, stores parameter change
+
+	name []byte
+
+	table TypeTable
+}
+
+// NewSubspace constructs a store with namestore
+func NewSubspace(cdc *codec.Codec, key sdk.StoreKey, tkey sdk.StoreKey, name string) (res Subspace) {
+	res = Subspace{
+		cdc:  cdc,
+		key:  key,
+		tkey: tkey,
+	}
+
+	namebz := []byte(name)
+	res.name = make([]byte, len(namebz), len(namebz)+extraKeyCap)
+	copy(res.name, namebz)
+	return
+}
+
+// WithTypeTable initializes TypeTable and returns modified Subspace
+func (s Subspace) WithTypeTable(table TypeTable) (res Subspace) {
+	if table == nil {
+		panic("SetTypeTable() called with nil TypeTable")
+	}
+	if s.table != nil {
+		panic("SetTypeTable() called on initialized Subspace")
+	}
+
+	res = Subspace{
+		cdc:   s.cdc,
+		key:   s.key,
+		tkey:  s.tkey,
+		name:  s.name,
+		table: table,
+	}
+
+	return
+}
+
+// Returns a KVStore identical with ctx.KVStore(s.key).Prefix()
+func (s Subspace) kvStore(ctx sdk.Context) sdk.KVStore {
+	// append here is safe, appends within a function won't cause
+	// weird side effects when its singlethreaded
+	return ctx.KVStore(s.key).Prefix(append(s.name, '/'))
+}
+
+// Returns a KVStore identical with ctx.TransientStore(s.tkey).Prefix()
+func (s Subspace) transientStore(ctx sdk.Context) sdk.KVStore {
+	// append here is safe, appends within a function won't cause
+	// weird side effects when its singlethreaded
+	return ctx.TransientStore(s.tkey).Prefix(append(s.name, '/'))
+}
+
+// Get parameter from store
+func (s Subspace) Get(ctx sdk.Context, key []byte, ptr interface{}) {
+	store := s.kvStore(ctx)
+	bz := store.Get(key)
+	err := s.cdc.UnmarshalJSON(bz, ptr)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// GetIfExists do not modify ptr if the stored parameter is nil
+func (s Subspace) GetIfExists(ctx sdk.Context, key []byte, ptr interface{}) {
+	store := s.kvStore(ctx)
+	bz := store.Get(key)
+	if bz == nil {
+		return
+	}
+	err := s.cdc.UnmarshalJSON(bz, ptr)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Get raw bytes of parameter from store
+func (s Subspace) GetRaw(ctx sdk.Context, key []byte) []byte {
+	store := s.kvStore(ctx)
+	return store.Get(key)
+}
+
+// Check if the parameter is set in the store
+func (s Subspace) Has(ctx sdk.Context, key []byte) bool {
+	store := s.kvStore(ctx)
+	return store.Has(key)
+}
+
+// Returns true if the parameter is set in the block
+func (s Subspace) Modified(ctx sdk.Context, key []byte) bool {
+	tstore := s.transientStore(ctx)
+	return tstore.Has(key)
+}
+
+// Set parameter, return error if stored parameter has different type from input
+// Also set to the transient store to record change
+func (s Subspace) Set(ctx sdk.Context, key []byte, param interface{}) {
+	store := s.kvStore(ctx)
+
+	ty, ok := s.table[string(key)]
+	if !ok {
+		panic("Parameter not registered")
+	}
+
+	pty := reflect.TypeOf(param)
+	if pty.Kind() == reflect.Ptr {
```

Comment:
I think this is a little counterintuitive - why not instead require that pointers be dereferenced before being passed to this function?

Reply:
`ParamSet.KeyValuePairs()` returns pairs of (ParamKey, PointerToTheStructField) since `Subspace.GetParamSet()` need to take the pointers. `Subspace.SetParamSet()` also using `KeyValuePairs` so it passes the pointers to `Subspace.Set()`

Also, I think it is better to keep in this way for the consistency with amino; amino also allows pointer argument and indirects it automatically.

Reply:
Haha OK, maybe my beef is with Amino then. Agreed on consistency.